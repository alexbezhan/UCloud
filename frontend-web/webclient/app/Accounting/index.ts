import Usage from "Accounting/Usage";

export {Usage};