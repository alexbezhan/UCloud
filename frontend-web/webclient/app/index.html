<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <title>UCloud</title>
</head>
<body>
    <noscript>
        JavaScript is required to run this page.
    </noscript>
    <div id="app"></div>
</body>
</html>
