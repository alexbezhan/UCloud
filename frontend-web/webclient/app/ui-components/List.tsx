import styled from "styled-components";
import Box from "./Box";

const childPadding = ({ childPadding }: { childPadding?: string | number }) =>
    childPadding ? { marginBottom: childPadding, marginTop: childPadding } : null;

const List = styled(Box) <{ fontSize?: string, childPadding?: string | number, bordered?: boolean }>`
    font-size: ${props => props.fontSize};
    & > * {
        ${props => props.bordered ? "border-bottom: 1px solid lightGrey;" : null }
        ${childPadding};
    }

    & > *:last-child {
        ${props => props.bordered ? "border-bottom: 0px;" : null}
    }
`;

List.defaultProps = {
    fontSize: "large",
    bordered: true
};

List.displayName = "List";

export default List;