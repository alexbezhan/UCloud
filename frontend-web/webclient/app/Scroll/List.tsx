import * as React from "react";
import {Button, Flex} from "ui-components";
import * as Heading from "ui-components/Heading";
import {ScrollRequest, ScrollResult, ScrollSize} from "./Types";

interface ListProps<Item, OffsetType> {
    scroll?: ScrollResult<Item, OffsetType>;
    scrollSize?: ScrollSize;

    frame?: (containerRef, children: React.ReactNode) => React.ReactNode;
    renderer: (props: {item: Item}) => JSX.Element | null;
    onNextScrollRequested: (request: ScrollRequest<OffsetType>) => void;
    spacer?: (height: number) => React.ReactNode;
    loading: boolean;

    customEmptyPage?: React.ReactNode;
}

interface ListState {
    firstVisibleElement?: number;
    lastVisibleElement?: number;
    spacingRequired?: number;
    postSpacingRequired?: number;
}

export class List<Item, OffsetType> extends React.Component<ListProps<Item, OffsetType>, ListState> {
    private eventListener: (e: UIEvent) => void;
    private container = React.createRef<HTMLElement>();
    private recordedHeights: number[] = [];
    private topOffset: number = -1;
    private averageComponentSize: number = -1;

    private get scrollOrDefault(): ScrollResult<Item, OffsetType> {
        return this.props.scroll || {endOfScroll: false, nextOffset: null, items: []};
    }

    private get scrollSizeOrDefault(): ScrollSize {
        return this.props.scrollSize || 50;
    }

    constructor(props: ListProps<Item, OffsetType>) {
        super(props);
        this.state = {};
    }

    public shouldComponentUpdate(nextProps: ListProps<Item, OffsetType>, nextState: ListState): boolean {
        return this.state !== nextState || this.props !== nextProps;
    }

    public getSnapshotBeforeUpdate(prevProps: Readonly<ListProps<Item, OffsetType>>): null {
        const scroll = prevProps.scroll || {endOfScroll: false, nextOffset: null, items: []};
        const nextScrollOrDefault = this.scrollOrDefault;

        this.topOffset = -1;

        if (nextScrollOrDefault.items.length < scroll.items.length) {
            this.recordedHeights = Array.from({length: scroll.items.length}, () => 0);
        } else {
            // TODO This is not always correct (but often is)
            const missingEntries = nextScrollOrDefault.items.length - scroll.items.length;
            this.recordedHeights = this.recordedHeights.concat(Array.from({length: missingEntries}, () => 0));
        }
        return null;
    }

    public componentDidMount(): void {
        // Attach scroll listener
        this.eventListener = e => {
            // Request more when we reach the end of the page
            if ((window.innerHeight + window.pageYOffset) >= document.body.offsetHeight - 200) {
                this.requestMore(false);
            }

            // Calculate the top and bottom of the scroll component. The topOffset contains how many pixels from the to
            // of the document we are offset by.
            if (this.topOffset === -1) {
                const container = this.container.current;
                if (container !== null) {
                    const bodyRect = document.body.getBoundingClientRect();
                    const containerRect = container.getBoundingClientRect();
                    this.topOffset = containerRect.top - bodyRect.top;
                } else {
                    this.topOffset = 0;
                }
            }

            const currentTop = window.scrollY;
            const currentBottom = currentTop + window.innerHeight - this.topOffset;

            // In this code we determine:
            //  - Which elements should be visible (firstVisibleElement and lastVisibleElement)
            //  - How much spacing to add before and after the visible elements (spacingRequired and
            //    postSpacingRequired)

            // The 'heights' property contains the previously recorded heights of components. After the scroller
            // updates we record the heights of the rendered children.
            const heights = this.recordedHeights;

            // The 'sum' is used for counting amount of spacing required. This is used for pre- and post-spacing.
            // We reset the property after having found the lastVisibleElement.
            //
            // Once the sum gets bigger than the currentTop we know that we have found the firstVisibleElement.
            // Similarly, when we get bigger than currentBottom we have found the lastVisibleElement.
            let sum = 0;

            let firstVisibleElement: number | undefined;
            let lastVisibleElement: number | undefined;
            let spacingRequired: number | undefined;
            let postSpacingRequired: number | undefined;

            for (let i = 0; i < heights.length; i++) {
                sum += heights[i];
                if (sum >= currentTop && firstVisibleElement === undefined) {
                    firstVisibleElement = i;
                    if (i > 0) {
                        spacingRequired = sum - heights[i];
                    }
                }

                if (sum >= currentBottom && lastVisibleElement === undefined) {
                    lastVisibleElement = i + 1;
                    sum = 0; // Start counting for postSpacingRequired
                }
            }

            if (lastVisibleElement !== undefined && sum > 0) {
                postSpacingRequired = sum;
            }

            // We only update the state if the visible elements have changed.
            // TODO(Dan): I assume that this is for performance reasons, but I don't actually remember why.
            const state = this.state;
            if (firstVisibleElement !== state.firstVisibleElement || lastVisibleElement !== state.lastVisibleElement) {
                this.setState(() => ({
                    firstVisibleElement,
                    lastVisibleElement,
                    spacingRequired,
                    postSpacingRequired
                }));
            }
        };

        window.addEventListener("scroll", this.eventListener);
    }

    public componentWillUnmount(): void {
        window.removeEventListener("scroll", this.eventListener);
    }

    public componentDidUpdate(): void {
        const container = this.container.current;
        if (container !== null) {
            const offset = this.state.firstVisibleElement || 0;
            // Skip the spacer
            const initialElement = this.state.spacingRequired !== undefined ? 1 : 0;
            const elementsToSkip = this.state.postSpacingRequired !== undefined ? 1 : 0;

            for (let i = initialElement; i < container.children.length - elementsToSkip; i++) {
                const child = container.children[i];
                this.recordedHeights[offset + i - initialElement] = child.getBoundingClientRect().height;
            }

            if (this.averageComponentSize === -1) {
                let sum = 0;
                let elementsToCount = 0;
                const recorded = this.recordedHeights;
                for (let i = 0; i < recorded.length; i++) {
                    if (recorded[i] > 0) {
                        sum += recorded[i];
                        elementsToCount++;
                    }
                }

                if (elementsToCount > 0) {
                    this.averageComponentSize = sum / elementsToCount;
                }
            }
        }
    }

    public render(): JSX.Element {
        return (
            <>
                {this.renderBody()}
                {this.renderLoadingButton()}
            </>
        );
    }

    private renderBody(): React.ReactNode {
        const {props} = this;
        if (props.loading && (props.scroll === undefined || props.scroll.items.length === 0)) {
            return null;
        } else {
            if (props.scroll === undefined || props.scroll.items.length === 0) {
                if (!props.customEmptyPage) {
                    return (
                        <div>
                            <Heading.h2>No results.</Heading.h2>
                        </div>
                    );
                } else {
                    return props.customEmptyPage;
                }
            } else {
                const children = this.renderEntries();
                const frame = props.frame;
                if (frame !== undefined) {
                    return frame(this.container, children);
                } else {
                    return <>{children}</>;
                }
            }
        }
    }

    private renderEntries(): React.ReactNode {
        const items = this.scrollOrDefault.items;
        const containerRef = this.container;
        const {renderer, spacer} = this.props;
        const {postSpacingRequired, spacingRequired, firstVisibleElement, lastVisibleElement} = this.state;

        return (
            <ListBody
                containerRef={containerRef}
                items={items}
                firstVisibleElement={firstVisibleElement}
                lastVisibleElement={lastVisibleElement}
                spacingRequired={spacingRequired}
                postSpacingRequired={postSpacingRequired}
                renderer={renderer}
                spacer={spacer}
            />
        );
    }

    private renderLoadingButton(): React.ReactNode {
        const {loading} = this.props;

        return (
            <Flex justifyContent={"center"}>
                <Button
                    onClick={() => this.requestMore(true)}
                    disabled={loading}
                >
                    Load more
                </Button>
            </Flex>
        );
    }

    private requestMore(alwaysLoadMore: boolean): void {
        const {loading, onNextScrollRequested} = this.props;
        const scroll = this.scrollOrDefault;
        const size = this.scrollSizeOrDefault;

        if (!loading && (!scroll.endOfScroll || alwaysLoadMore)) {
            onNextScrollRequested({
                offset: scroll.nextOffset,
                scrollSize: size
            });
        }
    }
}

interface ListBodyProps {
    spacingRequired?: number;
    postSpacingRequired?: number;
    firstVisibleElement?: number;
    lastVisibleElement?: number;
    containerRef?: React.RefObject<any>;
    items: any[];
    renderer: (props: {item: any}) => JSX.Element | null;
    spacer?: (height: number) => React.ReactNode;
}

class ListBody extends React.Component<ListBodyProps> {
    public shouldComponentUpdate(nextProps: ListBodyProps): boolean {
        const {props} = this;
        return props.spacingRequired !== nextProps.spacingRequired ||
            props.postSpacingRequired !== nextProps.postSpacingRequired ||
            props.firstVisibleElement !== nextProps.firstVisibleElement ||
            props.lastVisibleElement !== nextProps.lastVisibleElement ||
            props.items.length !== nextProps.items.length;
    }

    public render(): JSX.Element {
        const {
            spacingRequired, postSpacingRequired, firstVisibleElement, lastVisibleElement,
            containerRef, renderer, spacer, items
        } = this.props;

        const first = Math.max(0, firstVisibleElement || 0);
        const last = Math.min(items.length, lastVisibleElement !== undefined ? lastVisibleElement : items.length);

        const children: React.ReactNode[] = [];

        const spacerOrDefault = spacer || ((height: number) => <div key={`spacer${height}`} style={{height: `${height}px`}} />);
        if (spacingRequired !== undefined) {
            children.push(spacerOrDefault(spacingRequired));
        }

        for (let i = first; i < last; i++) {
            const Renderer = renderer;
            children.push(<Renderer key={i} item={items[i]} />);
        }

        if (postSpacingRequired !== undefined) {
            children.push(spacerOrDefault(postSpacingRequired));
        }

        if (containerRef !== undefined) {
            return <>{children}</>;
        } else {
            return <div ref={containerRef}>{children}</div>;
        }
    }
}
