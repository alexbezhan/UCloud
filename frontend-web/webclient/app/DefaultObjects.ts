import {SidebarOption, Page} from "Types";
import {Status} from "Navigation";
import {Analysis, DetailedApplicationSearchReduxState, RunsSortBy} from "Applications";
import {File, DetailedFileSearchReduxState, SortOrder} from "Files";
import {DashboardStateProps} from "Dashboard";
import {Notification} from "Notifications";
import {Upload} from "Uploader";
import {Activity, ActivityGroup, ActivityFilter} from "Activity";
import {Reducer} from "redux";
import {SimpleSearchStateProps} from "Search";
import * as ApplicationRedux from "Applications/Redux";
import * as AccountingRedux from "Accounting/Redux";
import {defaultAvatar} from "UserSettings/Avataaar";
import {SidebarPages} from "ui-components/Sidebar";
import {ScrollResult} from "Scroll/Types";
import * as ProjectRedux from "Project/Redux";

export const DefaultStatus: Status = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended."
};

export enum KeyCode {
    ENTER = 13,
    ESC = 27
}

export const emptyPage: Page<any> = {items: [], itemsPerPage: 25, itemsInTotal: 0, pageNumber: 0, pagesInTotal: 0};

export enum SensitivityLevel {
    "INHERIT" = "Inherit",
    "PRIVATE" = "Private",
    "CONFIDENTIAL" = "Confidential",
    "SENSITIVE" = "Sensitive"
}

export type Sensitivity = keyof typeof SensitivityLevel;

export enum SensitivityLevelMap {
    INHERIT = "INHERIT",
    PRIVATE = "PRIVATE",
    CONFIDENTIAL = "CONFIDENTIAL",
    SENSITIVE = "SENSITIVE"
}

export interface ComponentWithLoadingState {
    loading: boolean
    error?: string
}

export interface ComponentWithPage<T> extends ComponentWithLoadingState {
    page: Page<T>
}

export interface ComponentWithScroll<Item, OffsetType> extends ComponentWithLoadingState {
    scroll?: ScrollResult<Item, OffsetType>
}

export interface ResponsiveReduxObject {
    mediaType: string
    orientation: string
    lessThan: Record<string, boolean>
    greaterThan: Record<string, boolean>
    is: Record<string, boolean>
}

export const initResponsive = (): ResponsiveReduxObject => ({
    mediaType: "",
    orientation: "",
    lessThan: {},
    greaterThan: {},
    is: {},
});

export interface FilePreviewReduxState {
    file?: File
    error?: string
}

export interface FileInfoReduxObject {
    file?: File
    error?: string
    activity: Page<Activity>
    loading: boolean
}

export interface AnalysisReduxObject extends ComponentWithPage<Analysis> {
    sortBy: RunsSortBy
    sortOrder: SortOrder
}

export interface NotificationsReduxObject {
    redirectTo: string
    items: Notification[]
    loading: boolean
    error?: string
}

export interface StatusReduxObject {
    status: Status
    title: string
    page: SidebarPages
    loading: boolean
}

export interface SidebarReduxObject {
    pp: boolean
    options: SidebarOption[]
    kcCount: number
}

export interface HeaderSearchReduxObject {
    prioritizedSearch: HeaderSearchType
    refresh?: () => void
}

export interface RunApplicationReduxObject {

}

export type ActivityReduxObject = ComponentWithScroll<ActivityGroup, number> & ActivityFilter;

export type HeaderSearchType = "files" | "applications" | "projects";

export interface UploaderReduxObject {
    uploads: Upload[]
    visible: boolean
    path: string
    allowMultiple: boolean
    onFilesUploaded: (p: string) => void
    error?: string
    loading: boolean
}

interface LegacyReducers {
    dashboard?: Reducer<DashboardStateProps>
    uploader?: Reducer<UploaderReduxObject>
    status?: Reducer<StatusReduxObject>
    notifications?: Reducer<NotificationsReduxObject>
    analyses?: Reducer<AnalysisReduxObject>
    header?: Reducer<HeaderSearchReduxObject>
    sidebar?: Reducer<SidebarReduxObject>
    activity?: Reducer<ActivityReduxObject>
}

export type Reducers = LegacyReducers & ApplicationRedux.Reducers & AccountingRedux.Reducers;

/* FIXME */
interface LegacyReduxObject {
    dashboard: DashboardStateProps
    uploader: UploaderReduxObject
    status: StatusReduxObject,
    notifications: NotificationsReduxObject
    analyses: AnalysisReduxObject
    header: HeaderSearchReduxObject
    sidebar: SidebarReduxObject
    activity: ActivityReduxObject
    simpleSearch: SimpleSearchStateProps
    detailedFileSearch: DetailedFileSearchReduxState
    detailedApplicationSearch: DetailedApplicationSearchReduxState
    fileInfo: FileInfoReduxObject
    avatar: AvatarReduxObject
    filePreview: FilePreviewReduxState
    responsive?: ResponsiveReduxObject
    project: ProjectRedux.State
    loading?: boolean
}

export type ReduxObject =
    LegacyReduxObject &
    ApplicationRedux.Objects &
    AccountingRedux.Objects;

export const initActivity = (): ActivityReduxObject => ({
    loading: false
});

export const initNotifications = (): NotificationsReduxObject => ({
    items: [],
    loading: false,
    redirectTo: "",
    error: undefined
});

export const initHeader = (): HeaderSearchReduxObject => ({
    prioritizedSearch: "files"
});

export const initStatus = (): StatusReduxObject => ({
    status: DefaultStatus,
    title: "",
    page: SidebarPages.None,
    loading: false
});

export const initDashboard = (): DashboardStateProps => ({
    favoriteFiles: [],
    recentFiles: [],
    recentAnalyses: [],
    notifications: [],
    favoriteLoading: false,
    recentLoading: false,
    analysesLoading: false
});

export function initObject(homeFolder: string): ReduxObject {
    return {
        dashboard: initDashboard(),
        status: initStatus(),
        header: initHeader(),
        notifications: initNotifications(),
        analyses: initAnalyses(),
        sidebar: initSidebar(),
        uploader: initUploads(),
        activity: initActivity(),
        simpleSearch: initSimpleSearch(),
        detailedApplicationSearch: initApplicationsAdvancedSearch(),
        detailedFileSearch: initFilesDetailedSearch(),
        fileInfo: initFileInfo(),
        avatar: initAvatar(),
        filePreview: initFilePreview(),
        project: ProjectRedux.initialState,
        ...ApplicationRedux.init(),
        ...AccountingRedux.init(),
        responsive: undefined,
    }
}


export const initFilePreview = () => ({
    file: undefined,
    error: undefined
});

export type AvatarReduxObject = typeof defaultAvatar & { error?: string };
export const initAvatar = (): AvatarReduxObject => ({...defaultAvatar, error: undefined});

export const initSimpleSearch = (): SimpleSearchStateProps => ({
    files: emptyPage,
    filesLoading: false,
    applications: emptyPage,
    applicationsLoading: false,
    errors: [],
    search: "",
    applicationSearch: initApplicationsAdvancedSearch(),
    fileSearch: initFilesDetailedSearch()
});

export const initAnalyses = (): AnalysisReduxObject => ({
    page: emptyPage,
    loading: false,
    error: undefined,
    sortBy: RunsSortBy.createdAt,
    sortOrder: SortOrder.DESCENDING
});

export const initSidebar = (): SidebarReduxObject => ({
    pp: false,
    kcCount: 0,
    options: []
});

export const initUploads = (): UploaderReduxObject => ({
    path: "",
    uploads: [],
    visible: false,
    allowMultiple: false,
    error: undefined,
    onFilesUploaded: () => null,
    loading: false
});

export const initFileInfo = (): FileInfoReduxObject => ({
    activity: emptyPage,
    loading: false
});

export const initFilesDetailedSearch = (): DetailedFileSearchReduxState => ({
    hidden: true,
    allowFolders: true,
    allowFiles: true,
    fileName: "",
    extensions: new Set(),
    tags: new Set(),
    sensitivities: new Set(),
    createdBefore: undefined,
    createdAfter: undefined,
    modifiedBefore: undefined,
    modifiedAfter: undefined,
    error: undefined,
    loading: false
});

export const initApplicationsAdvancedSearch = (): DetailedApplicationSearchReduxState => ({
    error: undefined,
    loading: false,
    hidden: true,
    appName: "",
    appVersion: "",
    tags: ""
});
