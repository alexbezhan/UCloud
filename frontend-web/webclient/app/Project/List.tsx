import * as React from "react";
import {MainContainer} from "MainContainer/MainContainer";
import Button from "ui-components/Button";
import VerticalButtonGroup from "ui-components/VerticalButtonGroup";
import * as Pagination from "Pagination";
import {emptyPage} from "DefaultObjects";
import {useCloudAPI} from "Authentication/DataHook";
import {Page} from "Types";
import {listProjects, ListProjectsRequest, UserInProject} from "Project/index";
import Link from "ui-components/Link";
import Box from "ui-components/Box";
import {Dispatch} from "redux";
import {connect} from "react-redux";
import * as Heading from "ui-components/Heading";

const List: React.FunctionComponent<DispatchProps> = props => {
    const [response, setFetchParams, params] = useCloudAPI<Page<UserInProject>, ListProjectsRequest>(
        listProjects({page: 0, itemsPerPage: 50}),
        emptyPage
    );

    return <MainContainer
        headerSize={0}
        header={null}
        main={
            <Pagination.List
                page={response.data}
                pageRenderer={page => <>
                    {page.items.map(e =>
                        <ProjectSummary summary={e} setProject={props.setProject} key={e.id}/>
                    )}
                </>}
                loading={response.loading}
                onPageChanged={(newPage, page) => setFetchParams(listProjects({page: newPage, itemsPerPage: 50}))}
            />
        }
        sidebar={
            <VerticalButtonGroup>
                <Link to={"/projects/create"}><Button>Create</Button></Link>
                <Button color={"red"} onClick={e => props.setProject(undefined)}>Clear Project</Button>
            </VerticalButtonGroup>
        }
    />;
};

const ProjectSummary: React.FunctionComponent<{ summary: UserInProject } & DispatchProps> = props => {
    return <Box>
        <Heading.h3>{props.summary.title}</Heading.h3>
        <ul>
            <li>{props.summary.id}</li>
            <li>{props.summary.whoami.role}</li>
            <li><Link to={`/projects/view/${props.summary.id}`}>View</Link></li>
            <li><Link onClick={() => props.setProject(props.summary.id)}>Set as active</Link></li>
        </ul>
    </Box>;
};

interface DispatchProps {
    setProject: (id?: string) => void
}

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => ({
    setProject: (id?: string) => dispatch({type: "SET_PROJECT", project: id})
});

export default connect(null, mapDispatchToProps)(List);
