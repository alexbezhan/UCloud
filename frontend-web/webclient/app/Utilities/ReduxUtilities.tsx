import {createStore, combineReducers, Store, AnyAction, Action} from "redux";
import {composeWithDevTools} from 'redux-devtools-extension';
import {ReduxObject, initObject} from "DefaultObjects";
import {responsiveBP} from "ui-components/theme";
import {createResponsiveStateReducer} from "redux-responsive";
import {Cloud} from "Authentication/SDUCloudObject";
import {CONTEXT_SWITCH, USER_LOGOUT, USER_LOGIN} from "Navigation/Redux/HeaderReducer";

export function configureStore(initialObject: Partial<ReduxObject>, reducers, enhancers?): Store<ReduxObject, AnyAction> {
    const combinedReducers = combineReducers<ReduxObject, AnyAction>(reducers);
    const rootReducer = (state: ReduxObject, action: Action): ReduxObject => {
        if ([USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH].some(it => it === action.type)) {
            state = initObject(Cloud.homeFolder);
        }
        return combinedReducers(state, action);
    };
    return createStore<ReduxObject, AnyAction, {}, {}>(rootReducer, initialObject, composeWithDevTools(enhancers));
}

export const responsive = createResponsiveStateReducer(
    responsiveBP,
    {infinity: "xxl"}
);