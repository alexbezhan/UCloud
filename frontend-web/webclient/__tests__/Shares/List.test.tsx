import * as React from "react";
import List from "../../app/Shares/List";
import {create} from "react-test-renderer";
import {MemoryRouter} from "react-router";
import {configure} from "enzyme"
import * as Adapter from "enzyme-adapter-react-16";
import theme, {responsiveBP} from "../../app/ui-components/theme";
import {ThemeProvider} from "styled-components";
import "jest-styled-components";
import {configureStore} from "../../app/Utilities/ReduxUtilities";
import {initResponsive} from "../../app/DefaultObjects";
import {createResponsiveStateReducer} from "redux-responsive";
import {Provider} from "react-redux";

configure({adapter: new Adapter()});

const store = configureStore({responsive: initResponsive()}, {
    responsive: createResponsiveStateReducer(
        responsiveBP,
        {infinity: "xxl"}
    )
});

describe("Shares List", () => {
    test("Shares component", () => {
        expect(create(
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    <MemoryRouter>
                        <List/>
                    </MemoryRouter>
                </ThemeProvider>
            </Provider>)).toMatchSnapshot();
    });
});