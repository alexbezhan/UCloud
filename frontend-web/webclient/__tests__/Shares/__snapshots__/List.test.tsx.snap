// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shares List Shares component 1`] = `
.c0 {
  cursor: auto;
  margin-left: 68px;
  padding-top: 48px;
  padding-bottom: 14px;
  padding-left: 14px;
  padding-right: 0;
  background-color: #fff;
}

.c5 {
  cursor: auto;
  padding-top: 55px;
  padding-right: 14px;
}

.c7 {
  cursor: inherit;
  font-size: 24px;
  text-align: center;
  margin: 0;
  font-weight: 400;
}

.c6 {
  cursor: auto;
}

.c2 {
  cursor: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-bottom: 1px solid #f5f7f9;
  cursor: pointer;
}

.c3 {
  cursor: pointer;
  margin-right: 1em;
  border-bottom: 2px solid #006aff;
}

.c4 {
  cursor: pointer;
  margin-right: 1em;
}

.c1 {
  cursor: auto;
  padding-top: 14px;
  padding-bottom: 14px;
  padding-left: 82px;
  padding-right: 14px;
  width: 100%;
  height: 55px;
  background-color: #fff;
  position: absolute;
  top: 48px;
  left: 0;
  position: fixed;
  z-index: 50;
}

@media screen and (min-width:768px) and (max-width:1023px) {
  .c6 {
    display: none;
  }
}

@media screen and (min-width:1024px) and (max-width:1279px) {
  .c6 {
    display: none;
  }
}

@media screen and (min-width:1280px) {
  .c6 {
    display: none;
  }
}

<div
  className="c0"
  cursor="auto"
>
  <div
    className="c1"
    cursor="auto"
    height={55}
    width={1}
  >
    <div
      className="c2"
      cursor="inherit"
    >
      <div
        className="c3"
        cursor="pointer"
        onClick={[Function]}
        selected={true}
      >
        Shared with Me
      </div>
      <div
        className="c4"
        cursor="pointer"
        onClick={[Function]}
        selected={false}
      >
        Shared by Me
      </div>
    </div>
  </div>
  <div
    className="c5"
    cursor="auto"
  >
    <div
      className="c6"
      cursor="auto"
    />
    <h3
      className="c7"
      cursor="inherit"
      fontSize={4}
    >
      No shares
      <br />
      <small>
        Files shared will appear here.
      </small>
    </h3>
  </div>
</div>
`;
